### PR TITLE
[Duplicate of #5054]: Mock a Datasource for react using json-server

### LINKED ISSUE #4836: Mock  a Datasource for React using JSON Server
#### Proposed title of the article

Mock  a Datasource for React using JSON Server

#### Proposed article introduction

[Language] When building a React project, if it endpoints, developers end up needing some quick-to-use data sources. This provides test data to test the package if it works correctly.
Datasource mostly used may include DBMS systems such as MySQL, Postgres, SQL among many others. The problem with this approach is that during prototyping e.g. in Rapid Prototyping Approach if the application prototype is seemed to be not as expected it will be quickly discarded.
The extra setup and configurations will be both tedious and useless since they will be quickly discarded.

This produces a need to create a data source that can be quickly set up and discarded, yet it performs the same functionality as the DBMS.
It should also be small enough for fewer storage requirements and produce correct and accurate results as expected.
Such data sources problems can be solved by the use of the Json-Server library.

The Json-Server library is quick to install, configure, use and discard in a React application.
In this article, one will learn how to do the following on the package mentioned above in a React app.

#### KeyTakeaways

By the end, the following will be learned:

- What is the JSOn-Server library
- Set up the JSOn-Server library in a React application
- Connect the app to the library
- Use the library acts as a data source in place of a real DBMS
- Removing the dependency from the application and discarding the prototype

#### Pre-requisites

For the article, the following are required for easy article follow-up:

- A good React IDE or text editor set up on the machine
- A stable internet application
- React development environment already setup
- React development skills

#### Article quality

The article will be a span of broad content. It will begin with the basics before delving into more detailed but in-depth content on the library options.
This makes the user move from a beginner to expert level quickly.

#### References

- JSON Server [Documentations](https://www.npmjs.com/package/json-server)

#### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by blacklihgt**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #4836

### CODE REVIEW

**Comment by blacklihgt** [comment]
@srishilesh @hectorkambow I think I have finished my corrections and it's good to go... This is  a duplicate of the last pr #5054

**Comment by srishilesh** [comment]
Will complete the tech review by EOD

**Comment by srishilesh** [comment]
Jumping on this PR for tech review.

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+It has routes to access data items stored in the mock database file (JSON). Few examples are `GET /posts` for fetching all the posts, `PUT /posts/1` for updating the first post, or `DELETE /posts/1` for deleting the first post.
+
+The module allows other operations to be done on the database, such as:
+
+#### Filter
+Examples - `GET/posts?title=json-server&author=riro` or `GET/comments?author.name=riro`.
+
+These end-points filters `posts` with `title` of `json-server` and `author` name of `riro` and `comments` with the `author` name of `riro` respectively.
+
+An example is a `GET` request on `http://www.albana.com/posts?title=json-server&author=riro`.
```
Comment (srishilesh) [inline content/articles/mock-a-datasource-for-react-using-json-server/index.md:86]: This URL doesn't work

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+
+#### Pagination
+Examples - `GET/posts?_page=9&_limit=23`.
+
+This request will fetch posts from page 9, and the page limit is set to 23.
+
+An example is a GET request on `http://www.albana.com//posts?_page=9&_limit=23`.
```
Comment (srishilesh) [inline content/articles/mock-a-datasource-for-react-using-json-server/index.md:93]: All these example URLs doesn't work

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 1 above)
Comment (srishilesh) [inline content/articles/mock-a-datasource-for-react-using-json-server/index.md:86]: I've reviewed the article till here.

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+#### Operators
+e.g. for getting range (`_gte`, `lte`), excluding a value(`_ne`), filtering a value(`_like`). An example is a GET request on `http://www.albana.com/posts/4/comments?_ne=sad`. It excludes any comment with the value of 'sad' while `http://www.albana.com\author_like=chris` searches for authors with a name related to 'chris' (RegEx is used).
+#### Full-text search
+e.g. `GET/posts?q=tomcat`. This request searches for the value '**tomcat**' in all the records stored. An example is a GET request on `http://www.albana.com/posts?q=tomcat`.
+#### Relationships
+e.g., the inclusion of child resources(`_embed`), the inclusion of parent resources(`_expand`), to get or create nested resources. An example is `http://www.albana.com/comments/1?_expand=post` that includes parent resource by the name 'post'.
+#### Database
+e.g. `GET\db`. This GET request brings up all database items of the specified database. An example is `http://www.albana.com/customers_db`.
+#### Homepage
+e.g. `GET/`. Using the `./public` path, one can serve the application files in the public folder. An example is `http://www.albana.com/`.
```
Comment (srishilesh) [inline content/articles/mock-a-datasource-for-react-using-json-server/index.md:108]: Have a look at the above sub-topics and make changes accordingly

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+            <h1>{title}</h1>
+            {location.pathname === '/' && (
+                    <Button
+                            color={showAdd ? 'red' : 'green'}
+                            text={showAdd ? 'Close' : 'Add'}
+                            onClick={onAdd}
+                    />
+            )}
+          </header>
+  )
```
Comment (srishilesh) [inline content/articles/mock-a-datasource-for-react-using-json-server/index.md:177]: Are these formatted with 2 or 4 tab spaces each?

Possibly, you can choose 2

**Review by srishilesh — CHANGES_REQUESTED** [review]
@blacklihgt 

I've reviewed until line 88. There are very minor suggestions to be addressed before we could move on.

The example URLs doesn't seem to work, please correct them. The indentation looks good, but 2 tab spacing would look good, instead of 4.

Everything else looks good to me, except for some minor formatting and flow changes. The sooner you complete these requested changes, the sooner we can move for the final review.

Thanks.

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 4 above)
Comment (blacklihgt) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md:177]: ok

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 3 above)
Comment (blacklihgt) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md:108]: Let me take a look at the changes man

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 1 above)
Comment (blacklihgt) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md:86]: These were supposed to show the format of the URLs, not the actual URLs. I've corrected them

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 2 above)
Comment (blacklihgt) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md:93]: These were supposed to show the format of the URLs, not the actual URLs. I've corrected them

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 1 above)
Comment (blacklihgt) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md:86]: Ok, I hope you will be able to continue

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 3 above)
Comment (blacklihgt) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md:108]: please confirm if all is well

File: content/articles/mock-a-datasource-for-react-using-json-server/index.md (hunk 1 above)
Comment (srishilesh) [reply content/articles/mock-a-datasource-for-react-using-json-server/index.md]: It's better we have some mock server that responds to this request.

This URL would mislead the reader.

Probably, you can create a server that processes these requests with some sample data.

**Review by srishilesh — APPROVED** [review]
@blacklihgt 

Great work.

Make sure to go through my commits to understand the changes.

Moving for the final review.

Thanks.

**Comment by hectorkambow** [comment]
jumping on this one

**Comment by hectorkambow** [comment]
@blacklihgt 

Please be sure to go through the edits made by your peer reviewer and content moderator to improve future articles. 

The article was a bit difficult to read and some wording was off! 👀 

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published.

Our team of peer reviewers do not provide revisions services at this time in order to review them all in a timely manner. 
All revisions and edits should be completed on your own forked repo (as to not take up room in the queue). 

>Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

cc @srishilesh  Please do the same.

**Comment by hectorkambow** [comment]
@srishilesh any duplicate article that comes back should NOT require extensive revision.

**Comment by hectorkambow** [comment]
@blacklihgt can you reach out to me via Slack DM thank you

**Pull request merged**